### DISCUSSION TITLE
Using Webp in Web Commands (ImageSharp.Web)

### DISCUSSION THREAD

**Discussion by markadrake** [SixLabors/ImageSharp.Web#196] [Show and tell] [answered]
> Please have some patience with me – _I'm not a backend programmer._

It seems like **webp** support is right around the corner and I wanted to test some of it along with an Umbraco 9 build. When I try to issue the following command, the format request is ignored and I receive the original image:

```
{PATH_TO_YOUR_IMAGE}?format=webp
```

So I assume that **webp** it's not turned on by default and I have to adjust the configuration for `ImageSharp`? Also Umbraco may be coming into play here. I noticed they reference version 1.0.4. Everything works as expected with this version. Umbraco's implementation may need some updating or reconfiguration to support the latest and greatest.

Would someone validate my assumption and explain to me how I would go about registering / updating / correct configuring ImageSharp.Web?

I'll gladly donate something for your time. 

Thank you,

**Comment by brianpopow** [comment, marked as answer]
> So I assume that webp it's not turned on by default and I have to adjust the configuration for ImageSharp?

Webp is turned on by default, but its not yet released with a stable version (1.0.4 does not include webp yet). The first stable version of ImageSharp with webp will be version 2.0.0, which should not be far away.

If you really want to test its now, you will have to use the ImageSharp version from our nightly [MyGet feed](https://www.myget.org/feed/sixlabors/package/nuget/SixLabors.ImageSharp), version `2.0.0-alpha.0.142` for example. This needs to be added to the `ImageSharp.Web` instead of the `1.0.4`. But note that there are a few breaking changes with 2.0.0: For example the JpegEncoder option `Subsample` has changed to `ColorType`. You will need to change that in `JpegQualityWebProcessor`

**Reply by markadrake** [reply]
Thank you for that explanation @brianpopow.

When I installed ImageSharp.Web (1.0.5-alpha.0.2) I did not expect it to use the previous version of ImageSharp (1.0.4). 

I went ahead and installed ImageSharp (2.0.0-alpha.0.142).

I took a screenshot of my dependencies. It feels like magic, because years ago I know I'd be in some weird dependency / version hell. But it seems like dotnet recognizes the latest version installed. Hope I'm reading this correctly:

![image](https://user-images.githubusercontent.com/713355/147278872-d7ca69ea-fb8d-4704-9e05-3675ca7c1c31.png)

Now that I have the correct versions installed I'll continue down the rabbit hole here! Thanks for your response. I'm not sure I know how to go about updating Umbraco's default ImageSharp configuration but I will try to use the source as a guide. 

Once upgraded, many of the working web commands now fail. So there's probably some configuration updates needed including the one you mentioned for Jpeg. 

Thanks for your help.

**Reply by brianpopow** [reply]
I am not sure if that will work that way. The way i would do it:

- Clone `ImageSharp.Web`
- Add `ImageSharp.Web` to the Umbraco project
- Remove the `ImageSharp.Web` nuget reference from the Umbraco project and add a project reference instead (with the just added `ImageSharp.Web` project)
- Update the ImageSharp reference in Umbarco project and `ImageSharp.Web`
- Fix the build isuess

**Reply by markadrake** [reply]
Thanks Brian! I was able to follow this and get the project running with ease. Just a matter of actually importing the projects themselves, dealing with a few VS2022 caveats, and making the 1 change you suggested. I'm able to convert images to webp now. 

The resulting file size is a lot higher than expected when using the plain command:

```
{PATH_TO_YOUR_IMAGE}?format=webp
```

I'll continue to pick at the project. Thanks again for your help.

**Reply by JimBobSquarePants** [reply]
I've just created the first PR which will officially allow this functionality. https://github.com/SixLabors/ImageSharp.Web/pull/201

**Reply by JimBobSquarePants** [reply]
Re your file sizes. Using the sample app webp is giving me the best result 1kb less than jpeg with alpha transparency included!

![image](https://user-images.githubusercontent.com/385879/149780655-2aa0cbb4-7802-463e-93bb-b9573f5d4ee0.png)
